gpbft/gpbft.go

```diff
@@ -1201,30 +1224,21 @@ func (q *quorumState) FindStrongQuorumFor(key ChainKey) (QuorumResult, bool) {
 	panic("strong quorum exists but could not be found")
 }
 
-// Returns a list of the chains which have reached an agreeing strong quorum.
-// Chains are returned in descending length order.
-// This is appropriate for use in the QUALITY phase, where each participant
-// votes for every prefix of their preferred chain.
-// Panics if there are multiple chains of the same length with strong quorum
-// (signalling a violation of assumptions about the adversary).
-func (q *quorumState) ListStrongQuorumValues() []ECChain {
-	var withQuorum []ECChain
-	for key, cp := range q.chainSupport {
-		if cp.hasStrongQuorum {
-			withQuorum = append(withQuorum, q.chainSupport[key].chain)
-		}
+// FindStrongQuorumValueForLongestPrefixOf finds the longest prefix of preferred chain
+// which has strong quorum, or the base of preferred if no such prefix exists.
+func (q *quorumState) FindStrongQuorumValueForLongestPrefixOf(preferred ECChain) ECChain {
+	if q.HasStrongQuorumFor(preferred.Key()) {
+		return preferred
 	}
-	sort.Slice(withQuorum, func(i, j int) bool {
-		return len(withQuorum[i]) > len(withQuorum[j])
-	})
-	prevLength := 0
-	for _, v := range withQuorum {
-		if len(v) == prevLength {
-			panic(fmt.Sprintf("multiple chains of length %d with strong quorum", prevLength))
+	longestPrefix := preferred.BaseChain()
+	for _, cp := range q.chainSupport {
+		if cp.hasStrongQuorum &&
+			preferred.HasPrefix(cp.chain) &&
+			len(cp.chain) > len(longestPrefix) {
+			longestPrefix = cp.chain
 		}
-		prevLength = len(v)
 	}
```

Comment:
Don't we add chain support for all prefixes? We should iterate over all prefixes of `preferred` starting from the end, right?